python_modules/libraries/dagster-dbt/dagster_dbt/dbt_project.py

```diff
@@ -0,0 +1,212 @@
+import logging
+import os
+import shutil
+from pathlib import Path
+from typing import Optional, Sequence, Union
+
+from dagster._annotations import experimental, public
+from dagster._model import DagsterModel
+
+from .errors import DagsterDbtManifestNotFoundError, DagsterDbtProjectNotFoundError
+
+logger = logging.getLogger("dagster-dbt.artifacts")
+
+
+def using_dagster_dev() -> bool:
+    return bool(os.getenv("DAGSTER_IS_DEV_CLI"))
+
+
+def parse_on_load_opt_in() -> bool:
+    return bool(os.getenv("DAGSTER_DBT_PARSE_PROJECT_ON_LOAD"))
+
+
+@experimental
+class DbtManifestHandler:
+    """Life-cycle callbacks for dbt manifest represented by DbtProject."""
+
+    def on_load(self, project: "DbtProject"): ...
+    def on_prepare_for_deployment(
+        self,
+        project: "DbtProject",
+        *,
+        log_level: int = logging.INFO,
+    ): ...
+
+
+@experimental
+class DagsterDbtManifestHandler(DbtManifestHandler):
+    def __init__(
+        self,
+        generate_cli_args: Sequence[str] = ("parse", "--quiet"),
+    ):
+        """The default DbtManifestHandler, this handler provides an experience of:
+            * During development, reload the manifest at run time to pick up any changes.
+            * When deploying, expect a manifest that was created at build time to reduce start-up time.
+
+        Args:
+            generate_cli_args (Sequence[str]):
+                The arguments to pass to the dbt cli to generate a manifest.json.
+                Default: ("parse", "--quiet")
+        """
+        self._generate_cli_args = generate_cli_args
+
+    def on_load(self, project: "DbtProject"):
+        if using_dagster_dev() or parse_on_load_opt_in():
+            self.generate(project, log_level=logging.DEBUG)
+            if not project.manifest_path.exists():
+                raise DagsterDbtManifestNotFoundError(
+                    f"Did not find manifest.json at expected path {project.manifest_path} "
+                    f"after running '{self.generate.__qualname__}'. Ensure the implementation respects "
+                    "all DbtProject properties."
+                )
+
+    def on_prepare_for_deployment(self, project: "DbtProject", *, log_level: int = logging.INFO):
+        self.generate(project, log_level=log_level)
+
+    def generate(self, project: "DbtProject", *, log_level: int) -> Path:
+        from .core.resources_v2 import DbtCliResource
+
+        logger.log(
+            log_level,
+            f"Creating manifest.json for {project.project_dir} using {self._generate_cli_args}.",
+        )
+        manifest_path = (
+            DbtCliResource(project_dir=os.fspath(project.project_dir))
+            .cli(
+                self._generate_cli_args,
+                target_path=project.target_dir,
+            )
+            .wait()
+            .target_path.joinpath("manifest.json")
+        )
+        logger.log(log_level, f"Created {manifest_path}.")
+        return manifest_path
```

Comment:
things i like so far about this are
*  the dagster specific subclass `DagsterDbtManifestHandler` has the dagster specific set-up 
* these manifest management concerns are separated from the declarative `DbtProject`


but I dont feel attached to the "handler" naming and `on_` lifecycle approach - open to ideas on this one